src/ConnectedMode/CachingServerBranchProvider.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Core.Binding;
+
+namespace SonarLint.VisualStudio.ConnectedMode
+{
+    [Export(typeof(IStatefulServerBranchProvider))]
+    [Export(typeof(IBoundSolutionObserver))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    /// <summary>
+    /// Stateful decorator for <see cref="ServerBranchProvider"/> that updates when the bound solution changes
+    /// </summary>
+    internal class CachingServerBranchProvider : IStatefulServerBranchProvider, IBoundSolutionObserver
+    {
+        private readonly IServerBranchProvider branchProvider;
+
+        [ImportingConstructor]
+        public CachingServerBranchProvider(IServerBranchProvider serverBranchProvider)
+        {
+            this.branchProvider = serverBranchProvider;
+        }
+
+        public async Task<string> GetServerBranchNameAsync(CancellationToken token)
+        {
+            // TODO: caching
+
+            return await branchProvider.GetServerBranchNameAsync(token);
+        }
+
+        public void OnSolutionBindingChanged()
```

Comment:
This doesn't look right to me. We are introducing two ways to be notified when a solution binding has changed, either via implementing this delegate, or via importing ActiveSolutionBoundTracker and registering to its events. Working with the delegate doesn't allow the granularity of picking which solution event to register to (before or after). This is essentially means that if you want a Before then you should inherit from this, and if you want an After, you should register to the event.

I think it is cleaner to have just the events, with a Before event added to those that need it (which is this class).

ASBT already knows of IGitEvents (because it registers to HeadChanged), so it might as well call Refresh on it.

Reply:
I'll go with the event-based approach, although it doesn't really have anything in it's favour except consistency with the current (broken) approach.

When we look at fixing the eventing model, we should consider a callback-based approach:
* it's more loosely-coupled
* it's easier to test